File: app/src/main/java/com/example/cram_project/Spread.java
package com.example.cram_project;

import android.annotation.TargetApi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.TimeUnit;

public class Spread {
    
    //MAIN METHOD
    @TargetApi(24)
    public static Map<Date, HashMap<String, ArrayList<Workload>>> spread(ArrayList<Subject> subjects, ArrayList<Date> skip_dates) {

        //add exam dates to skip_dates list
        for(int i=0; i<subjects.size(); i++) {
            if(!skip_dates.contains(subjects.get(i).examDate)) {
                skip_dates.add(subjects.get(i).examDate);
            }
        }
        //spreading out workload process
        Map<Date, HashMap<String, ArrayList<Workload>>> calendar = new HashMap<>();
        for(Subject subject : subjects) {
            long total_days = TimeUnit.DAYS.convert((subject.endDate.getTime() - subject.startDate.getTime()), TimeUnit.MILLISECONDS);	//1 + (endDate - startDate)

            //if exam date within startDate & endDate, decrease study time (days) by 1 day
            for (Date date : skip_dates) {
                if(date.after(subject.startDate) && date.before(subject.endDate)) {
                    total_days -= 1;
                }
            }

            //evenly spread remaining workload between available days.
            int workPosition = 0;
            Date dateToStore;
            double skipValue = 0.0;
            double remainingSize = (double)subject.remainingWork.size()-1;

            //Finds evenly spread out values between 0.0 and total available days.
            //Start_date + the floor of each value is the date a workload is to be completed.
            for(double i=0.0; i<=total_days+0.0001; i += total_days/remainingSize) {
                dateToStore = incrementDateBy(subject.startDate, Math.floor(i)+skipValue);
                while(skip_dates.contains(dateToStore)) {       //if dateToBeAdded is to be skipped, go forward a day until you find free day
                    skipValue += 1.0;
                    dateToStore = incrementDateBy(subject.startDate, Math.floor(i)+skipValue);
                }
                if(dateToStore.after(subject.endDate)){         //if dateToBeAdded is AFTER end_date, go back a day until you find free day before end date
                    while(skip_dates.contains(dateToStore) || dateToStore.after(subject.endDate)){
                        dateToStore = decrementDateBy1(dateToStore);
                    }
                }

                calendar.putIfAbsent(dateToStore, new HashMap<String, ArrayList<Workload>>());
                calendar.get(dateToStore).putIfAbsent(subject.name, new ArrayList<Workload>());
                calendar.get(dateToStore).get(subject.name).add(subject.remainingWork.get(workPosition));
                workPosition += 1;
            }
        }

        //secondary spreading - taking difficulty weightings into account
        Map<Date, HashMap<String, ArrayList<Workload>>> sortedMap = new TreeMap<>(calendar);      //sorts dictionary into ascending order of dates

        int prevDifficulty = -99999;
        Date prevDate = null;
        for(Date date : sortedMap.keySet()) {
            int currentDifficulty = 0;
            for(String subject : sortedMap.get(date).keySet()) {
                for(Workload wl : sortedMap.get(date).get(subject)) {
                    currentDifficulty += wl.difficulty;
                }
            }
            //if previous date is weighted 2 or more in difficulty than current date, use wlToMove method to decide which workloads in the
            //previous date to move forward to the current date.
            if(prevDifficulty-currentDifficulty>1) {
                HashMap<String, ArrayList<Workload>> wlToMove = findWlToMove(sortedMap.get(prevDate), Math.floorDiv(prevDifficulty-currentDifficulty,2), date, subjects);

                for(String subj : wlToMove.keySet()) {
                    for(Workload wl : wlToMove.get(subj)) {
                        sortedMap.get(date).putIfAbsent(subj, new ArrayList<Workload>());
                        sortedMap.get(date).get(subj).add(wl);
                        sortedMap.get(prevDate).get(subj).remove(wl);
                        if(sortedMap.get(prevDate).get(subj).isEmpty()) {
                            sortedMap.get(prevDate).remove(subj);
                        }
                    }
                }
            }

            //if current date is weighted 2 or more in difficulty than previous date, use wlToMove method to decide which workloads in the
            //current date to move back to the previous date.
            else if(currentDifficulty-prevDifficulty>1 && prevDate != null) {
                HashMap<String, ArrayList<Workload>> wlToMove = findWlToMove(sortedMap.get(date), Math.floorDiv(currentDifficulty-prevDifficulty,2), prevDate, subjects);

                for (String subj : wlToMove.keySet()) {
                    for (Workload wl : wlToMove.get(subj)) {
                        sortedMap.get(prevDate).putIfAbsent(subj, new ArrayList<Workload>());
                        sortedMap.get(prevDate).get(subj).add(wl);
                        sortedMap.get(date).get(subj).remove(wl);
                        if (sortedMap.get(date).get(subj).isEmpty()) {
                            sortedMap.get(date).remove(subj);
                        }
                    }
                }
            }
            prevDifficulty = currentDifficulty;
            prevDate = date;
        }
        return sortedMap;
    }

    //helper method to increment a date by a number of days
    public static Date incrementDateBy(Date currentDate, double dta) {
        Calendar c = Calendar.getInstance();
        c.setTime(currentDate);
        c.add(Calendar.DATE, (int)dta);
        return c.getTime();
    }

    public static Date decrementDateBy1(Date currentDate){
        Calendar cal = Calendar.getInstance();
        cal.setTime (currentDate); // convert your date to Calendar object
        cal.add(Calendar.DATE, -1);
        return cal.getTime(); // again get back your date object
    }

    //NEXT STEP IS TO CREATE A REVERSE OF THIS METHOD!!!!!!!!!!!!!!!!!!!!!
    //helper method to find out which workloads to move from previous date to current date for equal spreading purposes.
    public static HashMap<String, ArrayList<Workload>> findWlToMove(HashMap<String, ArrayList<Workload>> workload, Integer difficulty, Date currentDate, ArrayList<Subject> subjects) {
        HashMap<String, ArrayList<Workload>> wlToMove = new HashMap<>();
        ArrayList<Workload> finalWl;
        Integer currentDifficultyTotal = 0;

        for(String subject : workload.keySet()) {
            //check if current subject being checked's end date comes AFTER the date workloads are being moved to
            boolean beforeEndAndAfterStart = false;
            for(Subject s : subjects){
                if (s.name == subject){
                    if((s.endDate == currentDate || s.endDate.after(currentDate)) && (s.startDate == currentDate || s.startDate.before(currentDate))){
                        beforeEndAndAfterStart = true;
                        break;
                    }
                }
            }
            if(beforeEndAndAfterStart) {
                //find the correct workloads to move forward
                for (Workload wl : workload.get(subject)) {
                    //if this workload is the exact difficulty weighting we're looking for, return that workload and we're done
                    if (wl.difficulty == difficulty) {
                        finalWl = new ArrayList<>(Arrays.asList(wl));
                        wlToMove = new HashMap<>();
                        wlToMove.put(subject, finalWl);
                        return wlToMove;
                    //if this workload is less than the difficulty we're looking for AND less than current found difficulty weighting -> add to workloads to be added list
                    } else if (wl.difficulty < difficulty && currentDifficultyTotal < difficulty - wl.difficulty) {
                        if (wlToMove.containsKey(subject)) {
                            wlToMove.get(subject).add(wl);
                        } else {
                            finalWl = new ArrayList<>(Arrays.asList(wl));
                            wlToMove.put(subject, finalWl);
                        }
                        currentDifficultyTotal += wl.difficulty;
                    }
                }
            }
        }
        return wlToMove;
    }
}

File: flutter_cram/flutter_app/lib/main/spread.dart
import 'dart:collection';

import 'package:flutter_app/database/databaseHelper.dart';
import 'package:flutter_app/main/futureArea.dart';
import 'package:flutter_app/model/subject.dart';
import 'package:flutter_app/model/workload.dart';

class Spread {
  spread(List<Subject> subjects, List<Workload> workloads, List<DateTime> skipDates) {

    //add exam dates to skipDates list
    for (int i = 0; i < subjects.length; i++) {
      if (!skipDates.contains(subjects[i].examDate)) {
        skipDates.add(DateTime.parse(subjects[i].examDate));
      }
    }

    //initially, spread workloads for subject between remaining study dates given
    Map<DateTime, Map<String, List<Workload>>> calendar = SplayTreeMap();
    int workPosition = 0;
    int initialWorkPos = 0;
    for (Subject subject in subjects) {
      DateTime startDate = DateTime.parse(subject.startDate);
      DateTime endDate = DateTime.parse(subject.endDate);

      //add in check to see if today's date is after start date here, if so make startDate today's date
      DateTime now = DateTime.now();
      if(startDate.isBefore(DateTime.now())){
        startDate = DateTime(now.year, now.month, now.day);
      }

      //if skip date within startDate and endDate, decrease studyTime (days) by 1 day
      int totalDays = endDate.difference(startDate).inDays;
      for (DateTime date in skipDates) {
        if (date.isAfter(startDate) && date.isBefore(endDate)) {
          totalDays -= 1;
        }
      }

      //evenly spread remaining workload between available days
      DateTime dateToStore;
      double skipValue = 0.0;
      double remainingWlNo = findRemainingWlNo(workloads, subject); //this integer should be calculated by using workloads database to find out which workloads have already been completed

      //Populate calendar with every day between start date and end date of subject EXCEPT SKIP DATES
      dateToStore = startDate;
      while((dateToStore.isBefore(endDate) || dateToStore == endDate)){
        if(!skipDates.contains(dateToStore)){
          calendar.putIfAbsent(dateToStore, () => Map<String, List<Workload>>());
        }
//        dateToStore = dateToStore.add(Duration(days: 1));
        dateToStore = DateTime(dateToStore.year, dateToStore.month, dateToStore.day+1);

      }

      //Finds evenly spread out values between 0.0 and total available days.
      //startDate + the floor of each value is the date a workload is to be completed
      for (double i = 0.0; (i <= totalDays + 0.001); i += totalDays / remainingWlNo) {
//        dateToStore = startDate.add(new Duration(days: (i + skipValue).floor()));
        dateToStore = DateTime(startDate.year, startDate.month, startDate.day+(i + skipValue).floor());
        while (skipDates.contains(dateToStore)) {
          skipValue += 1.0;
//          dateToStore = startDate.add(new Duration(days: (i + skipValue).floor()));
          dateToStore = DateTime(startDate.year, startDate.month, startDate.day+(i + skipValue).floor());

        }
        if (dateToStore.isAfter(endDate)) {
          while (skipDates.contains(dateToStore) || dateToStore.isAfter(endDate)) {
            dateToStore = DateTime(dateToStore.year, dateToStore.month, dateToStore.day-1);
          }
        }

        //find next available workload that is NOT completed and assign it to dateToStore.
        while(workloads[workPosition].complete==1){
          workPosition += 1;

          //prevent the loop from breaking
          if(workPosition>initialWorkPos+remainingWlNo){
            break;
          }
        }

        print("remaining wl no: " + remainingWlNo.toString());
        print("i: " + i.toString());
        print("work pos: " + workPosition.toString());
        print("max work pos: " + (initialWorkPos+remainingWlNo).toString());
        print("workload: " + workloads[workPosition].workloadName);
        print("date to store: " + dateToStore.toString());
        print("\n");

        workloads[workPosition].workloadDate = dateToStore.toString();

        //UPDATE DATE INTO WORKLOADS TABLE HERE
        updateWl(workloads[workPosition], DateTime.parse(workloads[workPosition].workloadDate));

        calendar.putIfAbsent(dateToStore, () => Map<String, List<Workload>>());
        calendar[dateToStore].putIfAbsent(subject.name, () => List<Workload>());
        calendar[dateToStore][subject.name].add(workloads[workPosition]);

        workPosition += 1;

        //prevent the loop from breaking
        if(workPosition>initialWorkPos+remainingWlNo){
          break;
        }
      }
      initialWorkPos = workPosition;
    }

//    //PRETTY PRINT FOR TESTING ------------------------------------------------------------------------
//    for (DateTime date in calendar.keys) {
//      int weight = 0;
//      print(date);
//      for (String subj in calendar[date].keys) {
//        for (Workload wl in calendar[date][subj]) {
//          print(subj + ": " + wl.workloadName + " - " + wl.workloadDate);
//          weight += wl.workloadDifficulty;
//        }
//      }
//      print("Weight: " + weight.toString());
//      print(
//          "_______________________________________________________________________");
//    }

    calendar = secondarySpread(calendar, subjects, workloads);
    calendar = secondarySpread(calendar, subjects, workloads);
    calendar = secondarySpread(calendar, subjects, workloads);
    calendar = secondarySpread(calendar, subjects, workloads);

    //Set date of complete workloads to NONE
    for(Workload wl in workloads){
      if(wl.complete == 1){
        updateWl(wl, null);
      }
    }
  }

  //secondary spreading - taking difficulty into account
  Map<DateTime, Map<String, List<Workload>>> secondarySpread(Map<DateTime, Map<String, List<Workload>>> calendar, List<Subject> subjects, List<Workload> workloads) {

    int prevDifficulty = 0;
    DateTime prevDate;
    for (DateTime date in calendar.keys) {
      int currentDifficulty = 0;
      for (String subject in calendar[date].keys) {
        for (Workload wl in calendar[date][subject]) {
          currentDifficulty += wl.workloadDifficulty;
        }
      }

      //if current date is weighted 2 or more in difficulty than previous date, use wlToMove method to decide which
      //workloads in the current date to move back to previous date
      if (currentDifficulty - prevDifficulty > 1 && prevDate != null) {
        Map<String, List<Workload>> wlToMove = findWlToMove(calendar[date], ((currentDifficulty - prevDifficulty) / 2).ceil(), prevDate, subjects);

//        //print test
//        for(String key in wlToMove.keys){
//          print(key + " moving back");
//          for(Workload wl in wlToMove[key]){
//            print(wl.workloadName);
//          }
//          print("_______________");
//        }

        for (String subj in wlToMove.keys) {
          for (Workload wl in wlToMove[subj]) {

            //UPDATE WORKLOAD DATABASE WITH DATE
            updateWl(wl, prevDate);

            //update workload date before adding to calendar
            wl.workloadDate = prevDate.toString();

            //update calendar
            calendar[prevDate].putIfAbsent(subj, () => List<Workload>());
            calendar[prevDate][subj].add(wl);
            calendar[date][subj].remove(wl);
            if (calendar[date][subj].isEmpty) {
              calendar[date].remove(subj);
            }
            prevDifficulty += wl.workloadDifficulty;
            currentDifficulty -= wl.workloadDifficulty;
          }
        }
      }

      //if previous date is weighted 2 or more in difficulty than current date, use wlToMove method to decide which
      //workloads in the previous date to move forward to current date
      if (prevDifficulty - currentDifficulty > 1) {
        Map<String, List<Workload>> wlToMove = findWlToMove(calendar[prevDate], ((prevDifficulty - currentDifficulty) / 2).ceil(), date, subjects);

//        //print test
//        for(String key in wlToMove.keys){
//          print(key + " moving forward");
//          for(Workload wl in wlToMove[key]){
//            print(wl.workloadName);
//          }
//          print("_______________");
//        }

        for (String subj in wlToMove.keys) {
          for (Workload wl in wlToMove[subj]) {
            //UPDATE DATE INTO WORKLOADS TABLE HERE
            updateWl(wl, date);

            //update workload date before adding to calendar
            wl.workloadDate = date.toString();

            //update calendar
            calendar[date].putIfAbsent(subj, () => List<Workload>());
            calendar[date][subj].add(wl);
            calendar[prevDate][subj].remove(wl);
            if (calendar[prevDate][subj].isEmpty) {
              calendar[prevDate].remove(subj);
            }
            currentDifficulty += wl.workloadDifficulty;
            prevDifficulty -= wl.workloadDifficulty;
          }
        }
      }
      prevDifficulty = currentDifficulty;
      prevDate = date;
    }
    return calendar;
  }

  //find workloads to be moved from one date to another
  Map<String, List<Workload>> findWlToMove(Map<String, List<Workload>> workload, int difficulty, DateTime currentDate, List<Subject> subjects) {
    Map<String, List<Workload>> wlToMove = Map();
    List<Workload> finalWl;
    int currentDifficultyTotal = 0;

    for (String subject in workload.keys) {
      //check if current subject being checked's endDate comes AFTER the date workloads are being moved to
      bool beforeEndAndAfterStart = false;
      for (Subject s in subjects) {
        DateTime startDate = DateTime.parse(s.startDate);
        DateTime endDate = DateTime.parse(s.endDate);

        //add in check to see if today's date is after start date here, if so make startDate today's date
        DateTime now = DateTime.now();
        if(startDate.isBefore(DateTime.now())){
          startDate = DateTime(now.year, now.month, now.day);
        }

        if (s.name == subject) {
          if ((endDate == currentDate || endDate.isAfter(currentDate)) && (startDate == currentDate || startDate.isBefore(currentDate))) {
            beforeEndAndAfterStart = true;
            break;
          }
        }
      }
      if (beforeEndAndAfterStart) {
        //find the correct workloads to move forward
        for (Workload wl in workload[subject]) {
          //if this workload is the exact difficulty weighting we're looking for, return that workload and we're done
          if (wl.workloadDifficulty == difficulty) {
            finalWl = [wl];
            wlToMove = Map();
            wlToMove[subject] = finalWl;
            return wlToMove;
          }
          else if (wl.workloadDifficulty < difficulty && currentDifficultyTotal < difficulty - wl.workloadDifficulty) {
            if (wlToMove.containsKey(subject)) {
              wlToMove[subject].add(wl);
            }
            else {finalWl = [wl];
              wlToMove[subject] = finalWl;
            }
            currentDifficultyTotal += wl.workloadDifficulty;
          }
        }
      }
    }
    return wlToMove;
  }

  //find remaining number of workloads yet to be completed
  double findRemainingWlNo(workloads, Subject subject){
    double i = 0.0;
    for(Workload wl in workloads){
      if(wl.complete == 0 && wl.subject == subject.name){
        i += 1.0;
      }
    }
    return i-1.0;
  }

  //UPDATE DATE INTO WORKLOADS TABLE HERE
  updateWl(Workload wl, DateTime date) {
    String newDate = "NONE";
    if(date != null){
      newDate = date.toString();
    }
    //update workloads database table
    var updatedWorkload = Workload(
      workloadID : wl.workloadID,
      subject : wl.subject,
      workloadName : wl.workloadName,
      workloadNumber : wl.workloadNumber,
      workloadDate : newDate,
      workloadDifficulty : wl.workloadDifficulty,
      complete : wl.complete,
    );
    DatabaseHelper.instance.updateWorkload(updatedWorkload);

    //update local workloads list
    for(Workload localWl in localWorkloads){
      if(localWl.workloadName == wl.workloadName){
        localWl.workloadDate = newDate;
        break;
      }
    }
  }
}
